### PR TITLE
[new-app] Morteza/save_store

### PR DESCRIPTION
**Pull request by morteza-binary**

### CODE REVIEW

File: src/javascript/app_2/Stores/base_store.js [hunk 1]
```diff
@@ -0,0 +1,139 @@
+import { action,
+    reaction,
+    toJS }               from 'mobx';
+import { isEmptyObject } from '../../_common/utility';
+
+/**
+ * BaseStore class is the base class for all defined stores in the application. It handle some stuff such as:
```
Comment (negarn) [inline src/javascript/app_2/Stores/base_store.js]: `handles`

File: src/javascript/app_2/Stores/base_store.js [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+ *  1. Creating snapshot object from the store.
+ *  2. Saving the store's snapshot in local/session storage and keep them sync.
```
Comment (negarn) [inline src/javascript/app_2/Stores/base_store.js]: `keeping them in sync`

File: src/javascript/app_2/Stores/base_store.js [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+ */
+export default class BaseStore {
+
+    /**
+     * An enume object to define LOCAL_STORAGE and SESSION_STORAGE
```
Comment (negarn) [inline src/javascript/app_2/Stores/base_store.js]: `enum`?

File: src/javascript/app_2/Stores/base_store.js [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+     */
+    static STORAGES = Object.freeze({
+        LOCAL_STORAGE  : Symbol('LOCAL_STORAGE'),
+        SESSION_STORAGE: Symbol('SESSION_STORAGE'),
+    });
+
+    /**
+     * Constructor of the base class that get properties name of child which should be saved in storages
```
Comment (negarn) [inline src/javascript/app_2/Stores/base_store.js]: `that gets the property names`?

File: src/javascript/app_2/Stores/base_store.js [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+     *
+     * @param {Object}   root_store - An object that contained the root store of the app.
```
Comment (negarn) [inline src/javascript/app_2/Stores/base_store.js]: `contains`?

File: src/javascript/app_2/Stores/base_store.js [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        this.local_storage_properties   = local_storage_properties || [];
+        this.session_storage_properties = session_storage_properties || [];
+        
+        this.setupReactionForLocalStorage();
+        this.setupReactionForSessionStorage();
+        this.retrieveFromStorage();
+    }
+
+    /**
+     * Returns an snapshot to the current store
```
Comment (negarn) [inline src/javascript/app_2/Stores/base_store.js]: `snapshot of the current store`?

File: src/javascript/app_2/Stores/base_store.js [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+     *
+     * @param {String[]} properties - A list of properties' names that should be in the snapshot.
+     *
+     * @return {Object} Returns a clone object for the store.
```
Comment (negarn) [inline src/javascript/app_2/Stores/base_store.js]: `cloned object of the store`?

File: src/javascript/app_2/Stores/base_store.js [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+                (result, p) => Object.assign(result, { [p]: snapshot[p]}),
+                {}
+            );
+        }
+
+        return snapshot;
+    }
+
+    /**
+     * Setups a reaction on properties which are mentioned in `local_storage_properties`
```
Comment (negarn) [inline src/javascript/app_2/Stores/base_store.js]: `sets up`

File: src/javascript/app_2/Stores/base_store.js [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+     *  and invoke `saveToStorage` when there are any changes on them.
```
Comment (negarn) [inline src/javascript/app_2/Stores/base_store.js]: `invokes`

File: src/javascript/app_2/Stores/base_store.js [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+        if (this.local_storage_properties.length) {
+            reaction(
+                () => this.local_storage_properties.map(i => this[i]),
+                () => this.saveToStorage(this.local_storage_properties, BaseStore.STORAGES.LOCAL_STORAGE)
+            );
+        }
+    }
+
+    /**
+     * Setups a reaction on properties which are mentioned in `session_storage_properties`
```
Comment (negarn) [inline src/javascript/app_2/Stores/base_store.js]: `sets up`

File: src/javascript/app_2/Stores/base_store.js [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+     *  and invoke `saveToStorage` when there are any changes on them.
```
Comment (negarn) [inline src/javascript/app_2/Stores/base_store.js]: `invokes`

File: src/javascript/app_2/Stores/base_store.js [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+            reaction(
+                () => this.local_storage_properties.map(i => this[i]),
+                () => this.saveToStorage(this.session_storage_properties, BaseStore.STORAGES.SESSION_STORAGE)
+            );
+        }
+    }
+
+    /**
+     * Saves the snapshot of the store includes properties that are mentioned in
+     *  first parameter in the storage that is mentioned in second parameters.
```
Comment (negarn) [inline src/javascript/app_2/Stores/base_store.js]: this whole sentence is confusing 😅 maybe:
`Store the passed properties in the particular storage and save a snapshot of it`

File: src/javascript/app_2/Stores/base_store.js [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+     *
+     * @param {String[]} properties - A list of name of properties of the store  which they should be saved on the storage.
```
Comment (negarn) [inline src/javascript/app_2/Stores/base_store.js]: `A list of the store's properties' names which should be saved in the storage`

File: src/javascript/app_2/Stores/base_store.js [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+     * @param {Symbol}   storage    - A symbol object that define the storage which the snapshot should stored on it.
```
Comment (negarn) [inline src/javascript/app_2/Stores/base_store.js]: `A symbol object that defines the storage which the snapshot should be stored in`

File: src/javascript/app_2/Stores/base_store.js [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+     *
+     */
+    saveToStorage (properties, storage) {
```
Comment (negarn) [inline src/javascript/app_2/Stores/base_store.js]: should there be space before `(`?

File: src/javascript/app_2/Stores/base_store.js [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+        const snapshot = JSON.stringify(this.getSnapshot(properties));
+
+        if ( storage === BaseStore.STORAGES.LOCAL_STORAGE) {
```
Comment (negarn) [inline src/javascript/app_2/Stores/base_store.js]: extra space after `(`

File: src/javascript/app_2/Stores/base_store.js (hunk 12 above)
Comment (morteza-binary) [reply src/javascript/app_2/Stores/base_store.js]: It doesn't store the properties in the particular storage. Actually, it gets a snapshot of the store and just removes properties that aren't mentioned in `properties` parameters from the snapshot and then save it to the particular storage.

Comment (negarn) [reply src/javascript/app_2/Stores/base_store.js]: I see! maybe summarise that and use it instead? 😝

File: src/javascript/app_2/Stores/base_store.js [hunk 17]
```diff
@@ -47,18 +48,18 @@ export default class BaseStore {
...
     }
 
     /**
-     * Returns an snapshot to the current store
+     * Returns an snapshot of the current store
      *
      * @param {String[]} properties - A list of properties' names that should be in the snapshot.
      *
-     * @return {Object} Returns a clone object for the store.
+     * @return {Object} Returns a cloned object for the store.
```
Comment (negarn) [inline src/javascript/app_2/Stores/base_store.js]: should it be `for` or `of`?

File: src/javascript/app_2/Stores/base_store.js (hunk 12 above)
Comment (morteza-binary) [reply src/javascript/app_2/Stores/base_store.js]: if you think the summarised one conveys the similar meaning, I'll change it.

Comment (negarn) [reply src/javascript/app_2/Stores/base_store.js]: I think the comment right now only mention `includes`, but your explanation is more like it uses those properties to decide which to remove, so maybe something like
`Removes properties that are not passed from the snapshot of the store and saves it to the passed storage`

Comment (negarn) [reply src/javascript/app_2/Stores/base_store.js]: if you can think of a clearer explanation please use that.

**Pull request merged**